Add blog outline: Open Source in the AI Age (#3)
ryscheng/www.raymondcheng.net: docusaurus/posts/2026-07-15-open-source-in-the-ai-age/index.md

---
slug: open-source-in-the-ai-age
title: Open Source will win in the AI Age
description: "Open source in the AI age: an opinionated answer to what open source becomes once AI writes the code, and how it caches society's progress and gives rise to super-powered maintainers."
authors: [ryscheng]
tags: [popular, thoughts, open-source]
draft: false
---

_An opinionated answer to what open source becomes once the AI writes the code:
how it will grow in importance as the caching layer for technological
breakthroughs, and why a sea of solo open source developers with superior taste
will out-innovate their closed counterparts._

<!-- truncate -->

## What is open source once AI writes the code?

What does **open source in the AI age** look like once the machines write most of
the code? Microsoft's CEO says [20-30% of the

> code](https://techcrunch.com/2025/04/29/microsoft-ceo-says-up-to-30-of-the-companys-code-was-written-by-ai/)

Author: This number is already quite outdated. We should remove the reference and just generically say that agents are writing larger fractions of codebases

**Depth.** A single developer with great taste, armed with AI, will out-build
entire teams. In the medium term, taste is the scarce input, and the people who
have it will produce more and better software than everyone else.

**Breadth.** Open source is how society caches its progress. Every published
library is a solved subproblem nobody has to solve again. In the long run, as
software teams go autonomous, that shared cache is the only thing that lets
exploration compound instead of being repeated in isolation.

(I've made [the parallel argument for data](/posts/network-intelligence)
elsewhere; here I'll focus on software.)

![Timeline: in the medium term, depth (the super-powered maintainer); in the long term, breadth (caching the search for what works); either way, open ecosystems out-innovate closed ones](./breadth-depth-timeline.svg)

## Depth: the super-powered maintainer

Give everyone the same coding agents, and taste becomes the only thing that
separates them. **A single developer with great taste will build a hundred great
libraries. A hundred developers with average taste will build a hundred mediocre
ones.** Same tools, wildly different output.

> Here is why. The prompt to the coding agent is the easy part. The hard part is

Author: This argument should be moved to the next section.

that millions of projects depend on, and he did it before agents could carry the
plumbing. Automate the boilerplate, the tests, the release chores, and the triage,
and the ceiling for one person with taste rises sharply.

Picture a small number of developers with exceptional taste, each running an army
of agents, trading ideas and setting the direction of whole ecosystems. Fewer
hands, far more leverage, better libraries.

Who pays these people is a real question, and I've [written
separately](/posts/prosperous-software) about funding the maintainers everything
depends on. The point here is simpler: **the super-powered maintainer is coming,
and taste is the superpower.**

## Breadth: how society caches its search for what works

Open source is a **cache**. Publishing a working library caches a solved
subproblem so nobody, anywhere, has to solve it again. The global commons of open
source is society's stored answer to _what works_.

Without that cache, every developer re-solves the same problems alone: the same
JSON parser, the same auth flow, the same date library. The same bugs. The same
incompatibilities. The same edge cases. Open source lets the whole world explore

> in parallel and keep the wins. Taste, the depth argument, is what keeps that

Author: ```suggestion
Without that cache, every developer re-solves the same problems alone: the same bugs, the same
incompatibilities, the same edge cases. Open source lets the whole world explore
in parallel and keep the wins. Taste, the depth argument, is what keeps that
```

---
slug: open-source-in-the-ai-age

> title: Open Source in the AI Age

Author: ```suggestion
title: Open Source will win in the AI Age
```

description: "Open source in the AI age: an opinionated answer to what open source becomes once AI writes the code, and how it caches society's progress and gives rise to super-powered maintainers."
authors: [ryscheng]
tags: [popular, thoughts, open-source]
draft: false
---

_An opinionated answer to what open source becomes once the AI writes the code:
how it caches society's progress, and why a single maintainer with taste will

> soon out-build entire teams._

Author: ```suggestion
how it will grow in importance as the caching layer for technological breakthroughs, and why a sea of solo open source developers with superior taste will out-innovate their closed counterparts._
```

<!-- truncate -->

## What is open source once AI writes the code?

What does **open source in the AI age** look like once the machines are writing
most of the code? Microsoft's CEO already says [20–30% of the
code](https://techcrunch.com/2025/04/29/microsoft-ceo-says-up-to-30-of-the-companys-code-was-written-by-ai/)
in the company's repositories is written by AI, and that fraction only climbs
from here. So the question sounds almost rhetorical: if a model can generate any

> library you need on demand, why would anyone bother publishing one?

Author: ```suggestion
library you need on demand, why do we even need open source libraries anymore?
```

> I think that intuition is exactly backwards.

Author: ```suggestion
_I think that intuition is exactly backwards._
```

Writing code used to be the expensive part. When it stops being expensive, the
bottleneck doesn't disappear. It moves. The hard question is no longer _can we
build it_ but _what is worth building, and how do we avoid everyone redoing the
same work_. That is precisely what open source is for.

So open source doesn't fade in the AI age. It matters more. My answer for _why_
comes in two parts, a breadth answer and a depth answer, and the interesting part
is how they trade off over time. (I've made [the parallel argument for

> data](/posts/network-intelligence) elsewhere; here I mean software.)

Author: ```suggestion
data](/posts/network-intelligence) elsewhere; here I'll focus on software.)
```

> ## Two answers, and when each one matters

Author: This section needs to be merged into the previous section, more concise and punchier arguments

Here are the two answers, each in a sentence.

**Breadth: open source is how society caches its progress and shares it — a
breadth-first search over what's worth building.** Every published library is a
solved subproblem that no one else has to solve again.

**Depth: a single maintainer with taste, amplified by AI, will soon out-build
entire teams.** One person with good judgment can produce hundreds of

> high-quality libraries, faster and better than a room full of engineers.

Author: ```suggestion
high-quality libraries, faster and better than a room full of mid-engineers, even if everyone is empowered with the same agents.
```

They aren't rival answers. They're the same phenomenon seen at two points on a
timeline:

- **Medium term — depth.** We still need human taste to steer, or the search
  degenerates into a wasteful crawl over every possible branch. The maintainer
  with taste keeps it tractable.
- **Long term — breadth.** Once we trust AI to design and experiment on its own,
  the volume of exploration explodes, and caching what works in the open is what
  lets the gains compound.

The next two sections take each answer in turn, starting with the one closer at
hand.

![Timeline: in the medium term, depth (the super-powered maintainer); in the long term, breadth (caching the search for what works); either way, open ecosystems out-innovate closed ones](./breadth-depth-timeline.svg)

## Depth: the super-powered maintainer

> Give a maintainer with genuine taste an army of tireless coding agents, and what

Author: This paragraph should make the point  that a single developer with great taste will make 100 great libraries better than 100 developers with average taste.

thousand variations; only taste picks the one worth publishing.

This isn't a fantasy that waits on future models. It's already visible in the
pre-AI world. Sindre Sorhus, one developer, [actively maintains over 1,000 npm
packages](https://sindresorhus.com/about), building blocks depended on by millions
of projects. He did that _before_ agents could carry the plumbing. Now imagine the
ceiling once the boilerplate, the tests, the release chores, and the issue triage
are largely automated.

The exciting version: a small number of 100x engineers, each with taste, trading
ideas and setting the direction of whole ecosystems. Fewer hands, more leverage,
better libraries.

There's an obvious open question, who pays these people, that I won't resolve
here; I've [written separately](/posts/prosperous-software) on how we might fund
the maintainers everything else depends on. The point for now is simpler: **the
super-powered maintainer is coming, and taste is their superpower.**

## Breadth: how society caches its search for what works

Think of open source as a **cache**. Publishing a working library is caching a
solved subproblem so that no one, anywhere, has to solve it again. The global

> commons of open source is society's memoized answer to _what works_.

Author: The prompt to the coding agent is the easy part. The hard part is knowing the package that is battle tested, works across various use cases, deployments, etc.

Why does that make it a _breadth_-first search? Because without a shared cache,

> every actor re-explores the same branches alone: the same JSON parser, the same

Author: the same bugs. The same incompatibilities, the same edge cases.

meets breadth: the maintainer's taste from the last section is what keeps that
search from being a blind, wasteful crawl.)

There's a durable economic reason caching wins, and it survives even as AI gets
cheap. The cost of producing software is falling fast, but it will never reach
zero, because model inference has a real, recurring cost. **As long as storing a
result is cheaper than recomputing it, caching pays.**

And it's already playing out in real time. Coding agents conserve effort exactly
the way a good engineer does: they reach for an existing dependency instead of
re-deriving or vendoring the code inline. The result is a visible surge in
open-source library installs.

![Combined monthly npm downloads for ten ubiquitous packages, rising from about 2.6 billion in mid-2024 to 5.9 billion by mid-2026, with the steepest climb in 2026](./npm-downloads.svg)
_Combined monthly npm downloads for ten ubiquitous packages (react, typescript, express, axios, and others). [Source: npm registry download API](https://api.npmjs.org/downloads/)_

Install volume for that basket is up roughly **69% year over year**, with the

> sharpest climb in 2026. I'll be honest about what the graph does and doesn't show:

Author: Don't say "I'll be honest". Talk in a more straight forward manner.

Author: Cut fluff

raw download counts include CI and automation, so this measures install
_activity_, not distinct developers, and the claim that _agents_ are driving it is
my interpretation, not something the chart proves. But the direction is
unmistakable.

Now extrapolate. When entire software teams go autonomous, designing and
experimenting without a human in the loop, the volume of exploration explodes.
Caching what works, in the open, becomes the only way that exploration compounds
across the world instead of being repeated behind a thousand closed doors.
**Depth is the medium-term face of this; breadth is the long-term one. They're

> the same thing.**

Author: How are they the same thing? They're different arguments. They support the same outcome, that open source out-innovates.

## As software gets cheaper, open becomes the default

There's a second force pushing the same direction, and it comes from the market,
not the technology.

As the cost of producing software collapses, people have started predicting the

> **death of SaaS** — a "SaaSpocalypse." The logic is simple: when a coding agent

Author: remove all em-dashes!

can rebuild your tool over a weekend, buyers grow far less willing to pay rent for
it. a16z devoted [an entire episode](https://a16z.com/podcast/anish-acharya-is-saas-dead-in-a-world-of-ai/)
to asking whether SaaS is dead in a world of AI; as they put it, "SaaS switching
costs are actually going down thanks to coding agents." The headlines are already
[declaring the collapse](https://www.forbes.com/sites/donmuir/2026/02/04/300-billion-evaporated-the-saaspocalypse-has-begun/).

> Follow the price. When software is cheap to produce and easy to substitute, its

Author: I really don't like this writing style. Be more direct and straight forward in explaining arguments.

market price drifts toward its falling cost, and tolerance for lock-in and markups
drops with it. Rent-seeking gets harder.

Now watch what that does to the release decision. **If a piece of software is hard
to monetize anyway, keeping it closed buys you almost nothing.** The friction of a
proprietary license is guarding a shrinking rent. The rational move flips toward
releasing it in the open, where you capture reputation, distribution, and
contributions instead.

So falling production cost doesn't threaten open source. It _feeds_ it. The same
force that ends easy SaaS rents pushes more code into the commons.

> I don't want to overstate this; it's a directional prediction, not a law.

Author: The writing style is getting worse. It is way too conversational and verbose here

engine.

So why do I take the bet? Two reasons.

First, the frontier has stayed reachable so far. The leading closed labs expose
their best models through APIs, and open-weight models have kept surprisingly close
behind. One mechanism there is distillation, training a smaller model on a larger
one's outputs, a decades-old and legitimate technique, if contested at the edges.
Contested enough that OpenAI has [formally accused
DeepSeek](https://www.bloomberg.com/news/articles/2026-02-12/openai-accuses-deepseek-of-distilling-us-models-to-gain-an-edge)
of distilling its models to catch up. That's an _allegation_, not an adjudicated
fact, but the fact that it's plausible tells you how hard the frontier is to keep
sealed.

Second, AI is its own precedent. The field advanced fastest precisely when
research flowed freely, and that same pressure toward openness now applies to the
models themselves.

Here's the honest failure mode, the thing that would prove me wrong: if frontier
capability concentrates in a few hands and _stays_ locked, with access metered,
weights sealed, and the gap widening, then the open ecosystem stalls and the

> closed players win. That's the bet. You can hold me to it.

Author: Okay this is really really bad.
